fem/gridfunc.hpp

```diff
@@ -476,21 +476,22 @@ class GridFunction : public Vector
                                              Array<int> &bdr_attr);
 
 
-   virtual double ComputeL2Error(Coefficient &exsol,
-                                 const IntegrationRule *irs[] = NULL) const
-   { return ComputeLpError(2.0, exsol, NULL, irs); }
-
    virtual double ComputeL2Error(Coefficient *exsol[],
-                                 const IntegrationRule *irs[] = NULL) const;
-
-   virtual double ComputeL2Error(VectorCoefficient &exsol,
                                  const IntegrationRule *irs[] = NULL,
                                  Array<int> *elems = NULL) const;
 
    /// Returns ||grad u_ex - grad u_h||_L2 in element ielem for H1 or L2 elements
    virtual double ComputeElementGradError(int ielem, VectorCoefficient *exgrad,
                                           const IntegrationRule *irs[] = NULL) const;
 
+   virtual double ComputeL2Error(Coefficient &exsol,
```

Comment:
Please add doxygen about `elems` to specify that it expects a marker instead of a list.

Reply:
Sure thing. Based on your previous comment, I will add a comment in the ComputeLp function. To avoid confusion though, I will wait until we address the input issue I mentioned in the previous comment.

Reply:
Sure thing.

Reply:
> Please add doxygen about `elems` to specify that it expects a marker instead of a list.

Is there a reason it is done this way and not using a list? I'm not sure what the use case for this is, but it would seem more efficient to only iterate over the elements in a list.

Reply:
Are you talking about something that is in mfem or exterior to it?

Reply:
A library user will have different ways of generating inputs to the interfaces. Maybe they will have to scan the whole list of elements to generate the active set before every call (as in your case it sounds like), or maybe they will already have the set and use it many times. Scanning everything and skipping in the latter case seems inefficient, especially if the set is small compared to the total size.

Reply:
So, you are suggesting I pass it only a list of elements that I want to compute the L2 error on, right? But to generate the list I would already need to use the markers to isolate those elements and pass it for L2 error computation. The element markers are computed separately. Once complete, the user passes it to the L2 error function where we compute the error on some of the elements depending on their marker.

Reply:
Let me put it a different way. What this interface needs is a representation of a subset of elements. I think a dense representation of the subset is preferable over a sparse representation, all other things being equal, because it consumes less memory, and uses fewer iterations to traverse.

It may be true in your case that you already have a sparse marker array, but there's no reason to believe this will be true in general for a user of the library. (I think it would be more common for a subset of elements to be described by a dense list of element numbers.)

For an example elsewhere in the library of the representation of a subset of elements in an API, the AMR interface takes a user-supplied subset of elements to refine. The interface receives a (dense) list of element numbers, not a (sparse) marker array, because it is more compact and does not waste cycles iterating over a possibly large number of no-ops.

For a little more motivation, suppose we had a multi-material setting, and want to compute errors in the elements of material N or on the boundaries of material N. If you only have 2 materials, then a marker strategy is probably fine. But suppose you had 50 materials, and wanted to compute the errors for boundaries of material N. Now you'd need to have 50 sparse marker arrays, most of which would be empty, and you'd be iterating over all of the elements 50 times each, mostly doing nothing. With the dense representation, you'd only visit each element once.

It is not a big deal, we could always add the dense list version later, it just seems like the natural choice for a general library API for describing a subset of elements, and is more consistent with other interface(s) where subsets of elements to process are supplied.

Reply:
Sorry for my late reply. I think we would still have one array of size the number of elements with the markers of interest. Meaning you would just mark the elements of interest according to some algorithm. So, I don't see why we would need N markers each of size of the number of elements. At least, what I had in mind was not to create N arrays for general problems.

Reply:
I don't want to hold the PR up so maybe it is best to proceed with the marker array, and we could add the list version later if needed.

But to clarify what I mean, when you say "But to generate the list I would already need to use the markers to isolate those elements and pass it for L2 error computation." This may be true for the way your code is organized. But I think it would be more common for the list to already exist. For example, a common data structure for material data is to have a ragged array enumerating elements of a given material. In that case, computing errors for material N would require filling out a full sized marked array, which adds no extra information and adds no-op iteration inside the library.

But as long as it's documenting as a marker array, probably best to proceed with this implementation for now.